### PR TITLE
adopt swift-collections 

### PR DESCRIPTION
**Pull request by tomerd**
motivation: replace TSC versions of ordered collections with the new ones from swift-collections

changes:
* pull in swift-collections as a dependency
* update bootstrap script
* adapt callsites to swift-collections
* reduce redundant imports, especially in tests across Basics, TSCBasic, TSCUtilities and SPMTestSupport

### CODE REVIEW

**Comment by tomerd** [comment]
this is the SwiftPM counterpart for to https://github.com/apple/swift-tools-support-core/pull/222

cc @WowbaggersLiquidLunch

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by WowbaggersLiquidLunch** [comment]
Although there is already #3533, I think this PR is better. I can close #3533 in favor of this.

File: Sources/Basics/Exports.swift [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+@_exported import OrderedCollections
+@_exported import TSCBasic
+// override TSC versions until deprecated
+// TODO: remove once TSC removes these
+public typealias OrderedSet = OrderedCollections.OrderedSet
+public typealias OrderedDictionary = OrderedCollections.OrderedDictionary
```
Comment (WowbaggersLiquidLunch) [inline Sources/Basics/Exports.swift:16]: This is a much cleaner solution than what I did in #3533. 

Just one question, though: Since there are only a few files that use `OrderedCollections`, is there any performance disadvantage for files that don't use `OrderedCollections` to import `Basics`?

File: Utilities/bootstrap [hunk 2]
```diff
@@ -174,6 +174,7 @@ def parse_global_args(args):
     args.swift_argument_parser_source_dir = os.path.join(args.project_root, "..", "swift-argument-parser")
     args.swift_driver_source_dir = os.path.join(args.project_root, "..", "swift-driver")
     args.swift_crypto_source_dir = os.path.join(args.project_root, "..", "swift-crypto")
+    args.swift_collections_source_dir = os.path.join(args.project_root, "..", "swift-collections")
```
Comment (WowbaggersLiquidLunch) [inline Utilities/bootstrap:177]: Changes in this file might conflict with #3582

File: Sources/Commands/SwiftBuildTool.swift [hunk 3]
```diff
@@ -13,7 +13,6 @@ import Basics
 import Build
 import PackageGraph
 import SPMBuildCore
-import TSCBasic
```
Comment (WowbaggersLiquidLunch) [inline Sources/Commands/SwiftBuildTool.swift:16]: By a rough count, there are about 10% of changes to `import`s not because of the changes to `Basics`. Would it be better to separate them into their own PR, to reduce the amount of files changed in this PR? Although, the PR probably still would contain changes to about 200 files.

**Review by WowbaggersLiquidLunch — APPROVED** [review]

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -164,6 +164,13 @@ For example, if the latest tag is 1.1.3:
 $> git clone https://github.com/apple/swift-crypto --branch 1.1.3
 ```
 
+5. Clone [swift-collections](https://github.com/apple/swift-collections) beside the SwiftPM directory and check out tag with the [latest version](https://github.com/apple/swift-collections/tags).
```
Comment (WowbaggersLiquidLunch) [inline CONTRIBUTING.md:167]: This should be list number 6, instead of 5

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -164,6 +164,13 @@ For example, if the latest tag is 1.1.3:
...
+
+For example, if the latest tag is 0.0.3:
+```bash
+$> git clone https://github.com/apple/swift-collections --branch 0.0.3
+```
+
```
Comment (WowbaggersLiquidLunch) [inline CONTRIBUTING.md:173]: Could the command examples be indented by 3 spaces so it's structurally more clear that they're part of the list items?

Also, could argument parser and swift crypto's versions in the commands be updated to 0.4.3 and 1.1.4, to align with what's in `Package.swift`?

File: Sources/Basics/Exports.swift (hunk 1 above)
Comment (tomerd) [reply Sources/Basics/Exports.swift:16]: oh, I totally missed #3533. sorry about that!

>  is there any performance disadvantage for files that don't use OrderedCollections to import Basics?

I dont believe so

File: Sources/Commands/SwiftBuildTool.swift (hunk 3 above)
Comment (tomerd) [reply Sources/Commands/SwiftBuildTool.swift:16]: exactly, this is why I ended up doing together since it would anyways be a huge PR

File: Sources/Basics/FileSystem+Extensions.swift [hunk 6]
```diff
@@ -8,7 +8,6 @@
  See http://swift.org/CONTRIBUTORS.txt for Swift project authors
  */
 
-import TSCBasic
```
Comment (neonichu) [inline Sources/Basics/FileSystem+Extensions.swift:11]: I don't think we should be removing the `TSCBasic` imports from all files since it currently has plenty of other types besides collections. We may want to remove the exports file prior not having `TSCBasic` anymore and then we have to put all those imports back again. If we have any files that only import `TSCBasic` for collections, we can remove the import from those of course.

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/Basics/FileSystem+Extensions.swift (hunk 6 above)
Comment (tomerd) [reply Sources/Basics/FileSystem+Extensions.swift:11]: what I did here was to replace the direct import on `TSCBasic` with an in import of `Basics` which currently re-exported by `TSCBasics`, the idea is that we can evolve `Basics` over time until we can eventually eliminate the need in `TSCBasic`

Comment (neonichu) [reply Sources/Basics/FileSystem+Extensions.swift:11]: I kinda feel like that could make it harder to remove `TSCBasic`? Since now every client of `Basics` will see the API of `TSCBasic` for an extended period of time. IMO, it would be good to get rid of the re-export quickly.

Comment (tomerd) [reply Sources/Basics/FileSystem+Extensions.swift:11]: is https://github.com/apple/swift-package-manager/pull/3595 better?

Comment (neonichu) [reply Sources/Basics/FileSystem+Extensions.swift:11]: Yah, I think I prefer that approach.
